src/Collections-Weak/WeakValueDictionary.class.st

```diff
@@ -16,10 +16,32 @@ Class {
 
 { #category : #adding }
 WeakValueDictionary >> add: anAssociation [
-	self at: anAssociation key put: anAssociation value.
+	"Set the value at key to be anObject.  If key is not found, create a new
+	entry for key and set is value to anObject. Answer anObject."
+
+	| index |
+	index := self findElementOrNil: anAssociation key.
+	(array at: index)
+		ifNil: [ self
+				atNewIndex: index
+				put:
+					((WeakValueAssociation key: anAssociation key value: anAssociation value asCollectionElement)
```

Comment:
Let's introduce #asWeakValueAssociation. So if it is already weak we will not create a copy.

Reply:
Indeed!